vllm/model_executor/layers/quantization/fp8_rocm.py

```diff
@@ -0,0 +1,303 @@
+        m = weight.shape[0]
+        n = x.shape[0]
+        k = x.shape[1]
+
+        algo = self._config._tuned.get((m, n, k))
+        if algo is None:
+            import os
+
+            if os.getenv("TUNE_FP8") == "1":
+                try:
+                    df = pd.read_csv("/projects/fp8_shapes.csv")
+                except:
+                    df = pd.DataFrame(columns=["M", "N", "K"])
+                df = pd.concat(
+                    [df, pd.DataFrame({"M": [m], "N": [n], "K": [k]})]
+                ).drop_duplicates()
+                df.to_csv("/projects/fp8_shapes.csv", index=False)
+            algo = 0
+        res = vllm_ops.fp8_gemm_16(x8, weight.t(), asf, wsf, int(algo))
+        return res
+
+    def apply_fp8_8(
+        self,
+        x: torch.Tensor,
+        weight: torch.Tensor,
+        asf: torch.Tensor,
+        wsf: torch.Tensor,
+        osf: torch.Tensor,
+        bias: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        assert not bias
+        x8 = torch.empty_like(x, dtype=torch.float8_e4m3fnuz)
+        vllm_ops.convert_fp8(x8, x, asf)
+        m = weight.shape[0]
+        n = x.shape[0]
+        k = x.shape[1]
+
+        algo = self._config._tuned.get((m, n, k))
+        if algo is None:
+            import os
+
+            if os.getenv("TUNE_FP8") == "1":
+                try:
+                    df = pd.read_csv("/projects/fp8_shapes.csv")
+                except:
+                    df = pd.DataFrame(columns=["M", "N", "K"])
+                df = pd.concat(
+                    [df, pd.DataFrame({"M": [m], "N": [n], "K": [k]})]
+                ).drop_duplicates()
+                df.to_csv("/projects/fp8_shapese.csv", index=False)
+            algo = 0
+
+        res = vllm_ops.fp8_gemm(x8, weight.t(), asf, wsf, osf, int(algo))
+        res16 = torch.empty_like(res, dtype=torch.float16)
+        vllm_ops.convert_fp8(res16, res, 1/osf)
+        return res16
+
+    def apply(
+        self,
+        layer: torch.nn.Module,
+        x: torch.Tensor,
+        bias: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        weight: torch.Tensor = layer.weight
+        if weight.dtype == torch.float8_e4m3fnuz:
+            asf: torch.Tensor = layer.activation_scaling_factor * 2
+            wsf: torch.Tensor = layer.weights_scaling_factor * 2
+            osf: torch.Tensor = layer.output_scaling_factor / 2
+
+            return self._config.gemm_method(self, x, weight, asf, wsf, osf)
+
+        return F.linear(x, weight, bias)
+    
+
+def _per_tensor_quantize(tensor: torch.Tensor,
+                        inv_scale: float) -> torch.Tensor:
+    finfo = torch.finfo(torch.float8_e4m3fnuz)
+    qweight = (tensor / inv_scale).clamp(min=finfo.min, max=finfo.max)
+    return qweight.to(torch.float8_e4m3fnuz)
+
+
+def _per_tensor_dequantize(tensor: torch.Tensor,
+                          inv_scale: float) -> torch.Tensor:
+    fake_qweight = tensor.to(torch.float16)
```

Comment:
Here is probably the only place where we can do that. This is in case we need to requantize merged fp8 (not used if the quantizer already outputs merged result). We requantize through fp16